### PR TITLE
Builder Enhancements

### PR DESCRIPTION
**Pull request by jeffreyladd**
# Summary
This enhances both the DiscordMessageBuilder and DiscordWebhookBuilder to have various changes

# Changes proposed
* Adds a Clear Method to both the DiscordMessageBuilder and DiscordWebhookBuilder so that they can be reused
* Removes the AddFile/WithFile overloads that takes a filePath
* Adds a param to the AddFile/WithFile Overloads that will allow for resetting the position of a File if it is requested.  This defaults to false
* Cleans up the internal methods of the Webhook Send so it will pass the builder the whole way through.
* Updates and adds the various tests that had to deal with this change.

### CODE REVIEW

**Comment by jeffreyladd** [comment]
This is ready for merge depending on the results of the poll that was conducted

**Comment by jeffreyladd** [comment]
With the results of the poll in, this is ready for Review and Merge

File: DSharpPlus/Entities/DiscordWebhookBuilder.cs [hunk 1]
```diff
@@ -151,7 +178,8 @@ public DiscordWebhookBuilder AddFile(string filename, Stream data)
         /// Adds the given files to send at the execution of the webhook.
         /// </summary>
         /// <param name="files">Dictionary of file name and file data.</param>
-        public DiscordWebhookBuilder AddFiles(Dictionary<string, Stream> files)
+        /// <param name="resetStreamPosition">Tells the API Client to reset the stream poition to what it was after the file is sent.</param>
```
Comment (Neuheit) [inline DSharpPlus/Entities/DiscordWebhookBuilder.cs:181]:
```suggestion
        /// <param name="resetStreamPosition">Tells the API Client to reset the stream position to what it was after the file is sent.</param>
```

File: DSharpPlus/Entities/DiscordWebhookBuilder.cs [hunk 2]
```diff
@@ -134,15 +135,41 @@ public DiscordWebhookBuilder AddEmbeds(IEnumerable<DiscordEmbed> embeds)
         /// </summary>
         /// <param name="filename">Name of the file.</param>
         /// <param name="data">File data.</param>
-        public DiscordWebhookBuilder AddFile(string filename, Stream data)
+        /// <param name="resetStreamPosition">Tells the API Client to reset the stream poition to what it was after the file is sent.</param>
```
[the remark this answers is not in the document]
Comment (jeffreyladd) [reply DSharpPlus/Entities/DiscordWebhookBuilder.cs]: This has been changed.

File: DSharpPlus/Entities/DiscordWebhookBuilder.cs (hunk 1 above)
Comment (jeffreyladd) [reply DSharpPlus/Entities/DiscordWebhookBuilder.cs]: This has been changed.

File: DSharpPlus/Entities/DiscordMessageBuilder.cs [hunk 3]
```diff
@@ -183,8 +170,9 @@ public DiscordMessageBuilder WithFile(string filePath)
         /// Sets if the message has files to be sent.
         /// </summary>
         /// <param name="files">The Files that should be sent.</param>
+        /// <param name="resetStreamPosition">Tells the API Client to reset the stream poition to what it was after the file is sent.</param>
```
[the remark this answers is not in the document]
Comment (jeffreyladd) [reply DSharpPlus/Entities/DiscordMessageBuilder.cs]: This has been changed.

File: DSharpPlus/Entities/DiscordMessageBuilder.cs [hunk 4]
```diff
@@ -145,36 +148,20 @@ public DiscordMessageBuilder WithFile(string fileName, Stream stream)
         /// Sets if the message has files to be sent.
         /// </summary>
         /// <param name="stream">The Stream to the file.</param>
+        /// <param name="resetStreamPosition">Tells the API Client to reset the stream poition to what it was after the file is sent.</param>
```
[the remark this answers is not in the document]
Comment (jeffreyladd) [reply DSharpPlus/Entities/DiscordMessageBuilder.cs]: This has been changed.

File: DSharpPlus/Entities/DiscordMessageBuilder.cs [hunk 5]
```diff
@@ -127,16 +126,20 @@ public DiscordMessageBuilder WithAllowedMentions(IEnumerable<IMention> allowedMe
         /// </summary>
         /// <param name="fileName">The fileName that the file should be sent as.</param>
         /// <param name="stream">The Stream to the file.</param>
+        /// <param name="resetStreamPosition">Tells the API Client to reset the stream poition to what it was after the file is sent.</param>
```
[the remark this answers is not in the document]
Comment (jeffreyladd) [reply DSharpPlus/Entities/DiscordMessageBuilder.cs]: This has been changed.

File: DSharpPlus/Entities/DiscordWebhookBuilder.cs [hunk 6]
```diff
@@ -187,5 +218,23 @@ public DiscordWebhookBuilder AddMentions(IEnumerable<IMention> mentions)
             this._mentions.AddRange(mentions);
             return this;
         }
+
+
+        public async Task<DiscordMessage> SendWebhook(DiscordWebhook webhook)
```
Comment (Neuheit) [inline DSharpPlus/Entities/DiscordWebhookBuilder.cs:223]: This should be renamed to `SendAsync()` to match consistency with the other builder.

Comment (jeffreyladd) [reply DSharpPlus/Entities/DiscordWebhookBuilder.cs]: done

**Review by Neuheit — APPROVED** [review]

**Pull request merged**